### PR TITLE
8287801: Fix test-bugs related to stress flags

### PR DESCRIPTION
**Pull request by eme64**
I recently ran many tests with additional stress flags. While there were a few bugs I found, most of the issues were test bugs.

I found a list of tests that are problematic with specific stress flags, I adjusted the tests so that they can now be run with the flag. Often I just fix the flag at the default value, so that setting it from the outside does not affect the test.

Below I explain for each test how and why I adjusted the test.

- test/hotspot/jtreg/compiler/arraycopy/TestArrayCopyNoInitDeopt.java
  - Problem Flags: `-XX:+UnlockExperimentalVMOptions -XX:PerMethodSpecTrapLimit=0 -XX:PerMethodTrapLimit=0`
    - Disabling traps by setting limit of traps to zero, means some optimistic optimizations are not made, and can therefore not lead to deoptimization. The test expects deoptimization due to traps, so we need to have them on.
- test/hotspot/jtreg/compiler/c2/cr7200264/TestDriver.java
  - used by TestSSE2IntVect.java and TestSSE4IntVect.java
- test/hotspot/jtreg/compiler/c2/cr7200264/TestSSE2IntVect.java
  - Problem Flags: `-XX:StressLongCountedLoop=2000000`
    - Test checks the IR, and if we convert loops to long loops, some operations will not show up anymore, the test fails. Disable StressLongCountedLoop.
- test/hotspot/jtreg/compiler/c2/cr7200264/TestSSE4IntVect.java
  - See TestSSE2IntVect.java
- test/hotspot/jtreg/compiler/c2/irTests/blackhole/BlackholeStoreStoreEATest.java
  - Problem Flags: `-XX:-UseTLAB`
    - No thread local allocation (TLAB) means IR is changed. Test checks for MemBarStoreStore, which is missing without TLAB. Solution: always have TLAB on.
- test/hotspot/jtreg/compiler/cha/AbstractRootMethod.java
  - Problem Flags: `-XX:+StressMethodHandleLinkerInlining`
    - Messes with recompilation, makes assert fail that expects recompilation. Must disable flag.
- test/hotspot/jtreg/compiler/cha/DefaultRootMethod.java
  - see AbstractRootMethod.java
- test/hotspot/jtreg/compiler/intrinsics/klass/CastNullCheckDroppingsTest.java
  - Problem Flags: `-XX:+UnlockExperimentalVMOptions -XX:PerMethodSpecTrapLimit=0 -XX:PerMethodTrapLimit=0`
    - Need traps, otherwise some optimistic optimisations are not made, and then they also are not trapped and deoptimized.
  - Problem Flags: `-XX:TypeProfileLevel=222`
    - Profiling also messes with optimizations / deoptimization.
  - Problem Flags: `-XX:+StressReflectiveCode`
    - Messes with types at allocation, which messes with optimizations.
  - Problem Flags: `-XX:-UncommonNullCast`
    - Is required for trapping in null checks.
  - Problem Flags: `-XX:+StressMethodHandleLinkerInlining`
    - Messes with inlining / optimization - turn it off.
- test/hotspot/jtreg/compiler/jvmci/compilerToVM/IsMatureVsReprofileTest.java
  - Problem Flags: `-XX:Tier4BackEdgeThreshold=1 -Xbatch -XX:-TieredCompilation`
  - Lead to OSR compilation in loop calling `testMethod`, which is expected to be compiled. But with the OSR compilation, that function is inlined, and never compiled. Solution was to make sure we only compile `testMethod`.
- test/hotspot/jtreg/compiler/jvmci/compilerToVM/ReprofileTest.java
  - Problem Flags: `-XX:TypeProfileLevel=222`
    - Changing profile flags messes with test, which assumes default behavior.
- test/hotspot/jtreg/compiler/profiling/TestTypeProfiling.java
  - Problem Flags: `-XX:+UnlockExperimentalVMOptions -XX:PerMethodSpecTrapLimit=0 -XX:PerMethodTrapLimit=0`
    - Need traps to check for optimistic optimizations.
- test/hotspot/jtreg/compiler/rangechecks/TestExplicitRangeChecks.java
  - Problem Flags: `-XX:+UnlockExperimentalVMOptions -XX:PerMethodSpecTrapLimit=0 -XX:PerMethodTrapLimit=0`
    - Need traps to check for optimistic optimizations.
- test/hotspot/jtreg/compiler/rangechecks/TestLongRangeCheck.java
  - Problem Flags: `-XX:+UnlockExperimentalVMOptions -XX:PerMethodSpecTrapLimit=0 -XX:PerMethodTrapLimit=0`
    - Need traps to check for optimistic optimizations.
- test/hotspot/jtreg/compiler/rangechecks/TestRangeCheckSmearing.java
  - Problem Flags: `-XX:TieredStopAtLevel=3 -XX:+StressLoopInvariantCodeMotion`
    - Test expects to be run at compilation tier 4 / C2, so must fix it at that in requirement.
- test/hotspot/jtreg/compiler/uncommontrap/Decompile.java
  - Problem Flags: `-XX:+UnlockExperimentalVMOptions -XX:PerMethodSpecTrapLimit=0 -XX:PerMethodTrapLimit=0 -XX:PerBytecodeTrapLimit=0`
  -   The tests if we trap and decompile after we call member functions of a class that we did not use before. If we disable traps, then internally it uses a virtual call, and no deoptimization is required - but the test expects trapping and deoptimization. Solution: set trap limits to default.
  - Problem Flags: `-XX:TypeProfileLevel=222`
  -   Changing profiling behavior also messes with deoptimization - disable it.
- test/hotspot/jtreg/compiler/uncommontrap/TestUnstableIfTrap.java
  - Problem Flags: `-XX:+UnlockExperimentalVMOptions -XX:PerMethodSpecTrapLimit=0 -XX:PerMethodTrapLimit=0`
    - Test expects traps, so we must ensure the limits are at default.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287801](https://bugs.openjdk.org/browse/JDK-8287801): Fix test-bugs related to stress flags

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [1c1da404](https://git.openjdk.org/jdk/pull/9186/files/1c1da4045ea2ae84605f401561d179b6ec3f7049)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9186/head:pull/9186` \
`$ git checkout pull/9186`

Update a local copy of the PR: \
`$ git checkout pull/9186` \
`$ git pull https://git.openjdk.org/jdk pull/9186/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9186`

View PR using the GUI difftool: \
`$ git pr show -t 9186`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9186.diff">https://git.openjdk.org/jdk/pull/9186.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/compiler/c2/irTests/blackhole/BlackholeStoreStoreEATest.java [hunk 1]
```diff
@@ -39,6 +39,7 @@ public class BlackholeStoreStoreEATest {
 
     public static void main(String[] args) {
         TestFramework.runWithFlags(
+            "-XX:+UseTLAB",
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/c2/irTests/blackhole/BlackholeStoreStoreEATest.java:42]: Just as a side note: As this example shows, treating `UseTLAB` as having no effect on the IR (i.e. on the IR framework whitelist) is incorrect. We should think about removing it from the whitelist at some point. But fixing this test like that is perfectly fine.

**Review by chhagedorn — APPROVED** [review]
Nice analysis and summary of the failures! The test fixes look good to me!

File: test/hotspot/jtreg/compiler/cha/AbstractRootMethod.java [hunk 2]
```diff
@@ -38,6 +38,7 @@
  *                   -XX:CompileCommand=compileonly,*::test -XX:CompileCommand=dontinline,*::test
  *                   -Xbatch -Xmixed -XX:+WhiteBoxAPI
  *                   -XX:-TieredCompilation
+ *                   -XX:-StressMethodHandleLinkerInlining
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/cha/AbstractRootMethod.java:41]: Wouldn't this fail with a product VM build because the flag is debug only?

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
Nice analysis. Please make sure to run all affected tests with a product VM build.

File: test/hotspot/jtreg/compiler/cha/AbstractRootMethod.java (hunk 2 above)
Comment (eme64) [reply test/hotspot/jtreg/compiler/cha/AbstractRootMethod.java:41]: Adding `-XX:+IgnoreUnrecognizedVMOptions` to 3 files, in other cases it is already there, so if a develop flag is used on a product build, that flag would simply be ignored.

**Review by TobiHartmann — APPROVED** [review]

**Pull request closed without merging**